jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java

```diff
@@ -180,61 +209,135 @@ public void succeeded(Connection connection)
                     {
                         if (LOG.isDebugEnabled())
                             LOG.debug("Connection {}/{} creation succeeded {}", total + 1, maxConnections, connection);
-                        connections.add(-1, 0);
-                        onCreated(connection);
-                        proceed();
+                        pendingConnectionCount.decrementAndGet();
```

Comment:
A thread past this statement has decremented the pending but not increased yet the pool size, so it's in a race with another thread entering `tryCreate()`.

Reply:
You're right, decrementing `pendingConnectionCount` before making the pool adopt the connection risks overshooting `maxConnections`. The only alternative is to first make the pool adopt the connection before decrementing `pendingConnectionCount`. There's a risk of not reaching `maxConnections` immediately if such race would occur, but later calls to `tryCreate` would eventually solve that so I think that's no big issue.